charms/candid/src/state.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for handling charm state."""
+
+import functools
+import json
+
+
+def requires_state_setter(func):
```

Comment:
Nice replacement for repeated `state.ready` checks.

Reply:
yes. but we have to be careful using it.. it is only useful when you're setting data in application's data bucket, which requires the unit to be leader.. if you're setting data in unit's data bucked, you don't need this check